pkg/controller/virtualnode.go

```diff
@@ -144,26 +145,45 @@ func (c *Controller) updateVNodeActive(vnode *appmeshv1beta1.VirtualNode, status
 }
 
 func (c *Controller) updateVNodeCondition(vnode *appmeshv1beta1.VirtualNode, conditionType appmeshv1beta1.VirtualNodeConditionType, status api.ConditionStatus) (*appmeshv1beta1.VirtualNode, error) {
-	now := metav1.Now()
 	condition := getVNodeCondition(conditionType, vnode.Status)
+	if condition.Status == status {
+		return nil, nil
+	}
+
+	now := metav1.Now()
 	if condition == (appmeshv1beta1.VirtualNodeCondition{}) {
 		// condition does not exist
-		newCondition := appmeshv1beta1.VirtualNodeCondition{
+		condition = appmeshv1beta1.VirtualNodeCondition{
 			Type:               conditionType,
 			Status:             status,
 			LastTransitionTime: &now,
 		}
-		vnode.Status.Conditions = append(vnode.Status.Conditions, newCondition)
-	} else if condition.Status == status {
-		// Already is set to status
-		return nil, nil
 	} else {
 		// condition exists and not set to status
 		condition.Status = status
 		condition.LastTransitionTime = &now
 	}
 
-	return c.meshclientset.AppmeshV1beta1().VirtualNodes(vnode.Namespace).UpdateStatus(vnode)
+	err := c.setVirtualNodeStatusCondition(vnode, condition)
+	return vnode, err
+}
+
+func (c *Controller) setVirtualNodeStatusCondition(vnode *appmeshv1beta1.VirtualNode, condition appmeshv1beta1.VirtualNodeCondition) error {
+	firstTry := true
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		var getErr error
+		if !firstTry {
+			vnode, getErr = c.meshclientset.AppmeshV1beta1().VirtualNodes(vnode.Namespace).Get(vnode.GetName(), metav1.GetOptions{})
+			if getErr != nil {
+				return getErr
+			}
+		}
+		vnodeCopy := vnode.DeepCopy()
+		vnodeCopy.Status.Conditions = append(vnode.Status.Conditions, condition)
```

Comment:
What happens here? On the first try we do a `Get` and overwrite the vnode parameter, but for the next retries we just do a deep-copy and keep appending the condition? 

Would it be possible to to the `Get` outside the retry loop? Why do we keep appending the same condition, can't we just do that once since only do one get? Same for `setVirtualNodeStatusCloudMapService()`

Reply:
It is inverse, on first try we do update directly and if update fails we try to get latest version and override. Ideally we should detect that previous update is not related to same field, but it is not easy to do in all cases. Also the status is derived from truth in AWS service in this case, so we should be okay.

Reply:
you are right about appending the condition, that is a bug. I will fix it. thanks

Reply:
Ah, thanks for the explanation. I guess the name `firstTry` with the negation confused me. Something like `updateVnode = false`, then setting it to `true` after the first try would be easier to read.